backend/matching_service/controllers/socketController.js

```diff
@@ -0,0 +1,95 @@
+const { QueueService } = require("../services/QueueService");
+const TIMEOUT_TIME = 30000;
+
+class SocketController {
+  constructor() {
+    this.socketMap = new Map();
+    this.queueService = new QueueService(this.socketMap);
+  }
+
+  handleConnection(socket) {
+    // socket.emit("connected", () =>
+    //   console.log("Connected to matching service")
+    // );
+
+    socket.on("startMatching", (data) =>
+      this.handleStartMatching(socket, data)
+    );
+
+    socket.on("cancelMatching", (uid) => this.handleCancelMatching(uid));
+  }
+
+  handleStartMatching(socket, { uid, difficulty, topic }) {
+    this.emitDoubleMatchingRequest(uid);
+    this.removeExistingConnection(uid); // Remove any existing connections for this user
+
+    // Add the socket to the map
+    const queueName = this.queueService.getQueueName(difficulty, topic);
+    this.socketMap.set(uid, { socket, queueName });
+
+    const sessionData = this.queueService.matchUser(queueName, uid);
+    if (sessionData) this.handleMatching(sessionData);
+
+    // Set a timeout for matching
+    setTimeout(() => this.handleTimeout(socket, uid), TIMEOUT_TIME);
+  }
+
+  handleCancelMatching(uid) {
+    this.removeExistingConnection(uid);
+  }
+
+  handleMatching(sessionData) {
+    const { prevUserSessionData, currUserSessionData } = sessionData;
+
+    const prevUserSocket = this.socketMap.get(prevUserSessionData.uid).socket;
+    const currUserSocket = this.socketMap.get(currUserSessionData.uid).socket;
+
+    prevUserSocket.emit("matched", prevUserSessionData);
+    currUserSocket.emit("matched", currUserSessionData);
+
+    prevUserSocket.disconnect();
+    currUserSocket.disconnect();
```

Comment:
Disconnection clean up is no longer triggered since `handleDisconnect()` is removed. Consider calling `removeExistingConnection()` instead to perform clean up.

Reply:
Great catch! I've updated it accordingly 😃